src/main/java/cz/sparko/boxitory/service/filesystem/FilesystemHashStore.java

```diff
@@ -0,0 +1,131 @@
+ */
+public class FilesystemHashStore implements HashStore {
+    private static final Logger LOG = LoggerFactory.getLogger(FilesystemHashStore.class);
+
+    /**
+     * Store checksum on filesystem, beside box file with appended extension derived from
+     * {@link HashAlgorithm}. When already exists, original file is not replaced.
+     * When file can't be created for any reason (e.g. write permissions), error is logged but no exception raised.
+     *
+     * @param box       absolute filesystem path to the box
+     * @param hash      calculated hash
+     * @param algorithm algorithm of hash
+     * @throws IllegalStateException when file provided by {@code box} does not exists
+     */
+    @Override
+    public void persist(String box, String hash, HashAlgorithm algorithm) {
+        File boxFile = checkBoxFileExists(box);
+
+        if (algorithm == HashAlgorithm.DISABLED) {
+            LOG.debug("Hash algorithm [{}]. Nothing to persist.", algorithm);
+            return;
+        }
+
+        try {
+            String boxHashFilename = box + algorithm.getFileExtension();
+            File hashFile = new File(boxHashFilename);
+            if (hashFile.exists()) {
+                LOG.trace("Hash file [{}] already exist. Not replacing!", boxHashFilename);
+                return;
+            }
+            boolean fileCreated = hashFile.createNewFile();
+            if (fileCreated) {
+                try (FileWriter writer = new FileWriter(hashFile)) {
+                    writer.write(createValidChecksumFilecontent(hash, boxFile));
+                    LOG.debug("Storing hash file [{}]", hashFile.getName());
+                }
+            } else {
+                LOG.debug("Hash file [{}] was not created. Dropping.");
+            }
+        } catch (IOException e) {
+            LOG.error("Something went wrong when persisting hash file.", e);
+        }
+    }
+
+    /**
+     * Reads checksum from filesystem, from file beside box file with extension derived from {@link HashAlgorithm}.
+     *
+     * @param box       absolute filesystem path to the box
+     * @param algorithm algorithm of hash
+     * @return hash when file found, {@link Optional#empty()} otherwise
+     * @throws IllegalStateException when file provided by {@code box} does not exists
+     */
+    @Override
+    public Optional<String> loadHash(String box, HashAlgorithm algorithm) {
+        File boxFile = checkBoxFileExists(box);
+
+        final Optional<String> hash;
+
+        File boxHashFile = new File(box + algorithm.getFileExtension());
+        if (boxHashFile.exists() && boxHashFile.isFile()) {
+            LOG.trace("Found hash file [{}] for box [{}]", boxHashFile.getAbsolutePath(), box);
+            try {
+                hash = Optional.of(readHashFromChecksumFile(boxHashFile, boxFile.getName()));
+                LOG.trace("Hash [{}] loaded from file [{}] for box [{}]", hash, boxHashFile.getAbsolutePath(), box);
```

Comment:
Is it good to call there trace? For me there should be info log.

Reply:
I made it debug. imho info is too strong. It would be logged for all box versions for every single request.